From 73443c39b3226a727d3e3c0cd7d7ebf6e1394603 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Thu, 04 Jul 2024 17:19:38 +0300
Subject: [PATCH] [FSW-2704] fs1: read switch manifest from cold flops

In case cold reset flops are set for FLOPS_UPD_PENDING
that means new firmware has been updated in manifest1.
In this case bootrom will try to boot manifest1 first
and treat manifest0 as secondary.

Change-Id: Icb941e4604510b16852b1d42223344b4cbfbbe88
---

diff --git a/app/boot_rom.c b/app/boot_rom.c
index a9d457a..417e24f 100644
--- a/app/boot_rom.c
+++ b/app/boot_rom.c
@@ -15,6 +15,7 @@
 	.recovery_flag = 0,
 	.validation_pass = 0,
 	.valid_manifest_idx = 0,
+	.fw_upd_switch_manifest = 0,
 	.efuse_read = 0,
 	.curr_img = 0,
 	.fake_fuse_en = false,
diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 5e46dc2..58801f9 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -16,6 +16,7 @@
 #include <fs1_reg_mask.h>
 #include <fs1_fw_if.h>
 #include <fs1_security.h>
+#include <hl_fw_upd_if.h>
 #include <status.h>
 #include <efuse.h>
 #include <hl_fw_upd_if.h>
@@ -130,6 +131,8 @@
 void boot_rom_read_cold_flops(void)
 {
 	struct fs1_cold_rst_data cold_rst_data;
+	enum fw_post_reset_cmd fw_upd_status =
+		regmap_readl(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_REG_3));
 
 	cold_rst_data.data = regmap_readl(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_REG_2));
 
@@ -139,6 +142,11 @@
 	boot_rom_global.efuse_read = cold_rst_data.efuse_read_flag;
 	boot_rom_global.boot_info_set_done = cold_rst_data.boot_info_set_done;
 	boot_rom_global.tpm_extend_done = cold_rst_data.tpm_validation_flag;
+
+	if (fw_upd_status == FLOPS_UPD_PENDING)
+		boot_rom_global.fw_upd_switch_manifest = 1;
+	else
+		boot_rom_global.fw_upd_switch_manifest = 0;
 }
 
 int boot_rom_mem_init(uint64_t addr, uint32_t size)
diff --git a/include/boot_rom.h b/include/boot_rom.h
index 3dcdf94..914db5a 100644
--- a/include/boot_rom.h
+++ b/include/boot_rom.h
@@ -89,6 +89,7 @@
 	uint8_t recovery_flag;
 	uint8_t validation_pass;
 	uint8_t valid_manifest_idx;
+	uint8_t fw_upd_switch_manifest;
 	uint8_t efuse_read;
 	uint8_t curr_img;
 	uint8_t die_id;
